File: test/wget_test.py
#!/usr/bin/env python

########################################################################
# wget_test.py: Test suite for wget.py
#
#  Description:
#  This test suite verifies the functionality of the wget.py script.
#  It uses mocking to simulate network requests and file operations
#  to avoid actual downloads during testing. The suite covers various
#  edge cases, such as normal downloads, HTTP errors, missing arguments,
#  file overwrites, and network timeouts, ensuring that the script behaves
#  as expected in different scenarios without accessing external resources
#  or modifying the real file system.
#
#  Author: id774 (More info: http://id774.net)
#  Source Code: https://github.com/id774/scripts
#  License: The GPL version 3, or LGPL version 3 (Dual License).
#  Contact: idnanashi@gmail.com
#
#  Version History:
#  v1.0 2025-01-10
#       Initial release. Test suite for wget.py script.
#
########################################################################

import os
import subprocess
import sys
import unittest
from unittest.mock import MagicMock, mock_open, patch

# Adjust the path to import script from the parent directory
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

try:
    import requests

    from wget import download_file, usage
    HAS_REQUESTS = True
except ImportError:
    HAS_REQUESTS = False


class TestWget(unittest.TestCase):
    """ Test suite for wget.py """

    def setUp(self):
        if not HAS_REQUESTS:
            self.skipTest("requests module is not installed")

    def test_usage_shows_help(self):
        script_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
        script_path = os.path.join(script_dir, 'wget.py')

        proc = subprocess.Popen(['python3', script_path, '-h'],
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE)
        out, err = proc.communicate()

        self.assertEqual(proc.returncode, 0)
        self.assertIn('Usage:', out.decode('utf-8'))

    @patch('wget.requests.get')
    @patch('builtins.open', new_callable=mock_open)
    def test_download_file(self, mock_open_file, mock_requests_get):
        """
        Test case: Normal file download.
        This test verifies that:
        - The file is correctly downloaded from the given URL.
        - The downloaded content is saved locally with the correct filename.
        """
        # Mock response content for the GET request
        mock_requests_get.return_value = MagicMock(status_code=200, content=b'Test content')

        # Call the download function
        download_file("http://example.com/testfile.txt")

        # Verify that the file is opened for writing in binary mode
        mock_open_file.assert_called_once_with("testfile.txt", 'wb')

        # Verify that the file's content is written correctly
        mock_open_file().write.assert_called_once_with(b'Test content')

    @patch('wget.requests.get')
    def test_http_error(self, mock_requests_get):
        """
        Test case: HTTP error during file download.
        This test verifies that:
        - HTTP errors such as 404 or 500 are properly raised as exceptions.
        """
        # Mock a RequestException to simulate an HTTP error
        mock_requests_get.side_effect = requests.exceptions.RequestException("HTTP Error")

        # Verify that the function raises the correct exception
        with self.assertRaises(requests.exceptions.RequestException):
            download_file("http://example.com/testfile.txt")

    @patch('wget.requests.get')
    def test_invalid_url(self, mock_requests_get):
        """
        Test case: Invalid URL provided.
        This test verifies that:
        - An invalid URL raises a MissingSchema exception.
        """
        # Mock a MissingSchema exception for invalid URLs
        mock_requests_get.side_effect = requests.exceptions.MissingSchema("Invalid URL")

        # Verify that the function raises the correct exception
        with self.assertRaises(requests.exceptions.MissingSchema):
            download_file("invalid-url")

    @patch('wget.requests.get')
    @patch('builtins.open', new_callable=mock_open)
    def test_file_overwrite(self, mock_open_file, mock_requests_get):
        """
        Test case: File overwrite.
        Verifies that existing files are overwritten with new content.
        """
        mock_requests_get.return_value.content = b'New content'
        mock_requests_get.return_value.status_code = 200

        download_file("http://example.com/testfile.txt")

        mock_open_file.assert_called_once_with("testfile.txt", 'wb')
        mock_open_file().write.assert_called_once_with(b'New content')

    @patch('wget.requests.get')
    def test_network_timeout(self, mock_requests_get):
        """
        Test case: Network timeout during download.
        Ensures the script handles timeouts gracefully.
        """
        mock_requests_get.side_effect = requests.exceptions.Timeout

        with self.assertRaises(requests.exceptions.Timeout):
            download_file("http://example.com/timeoutfile.txt")


if __name__ == '__main__':
    unittest.main()

File: test/wget_test.rb
#!/usr/bin/env ruby

########################################################################
# wget_test.rb: Test suite for wget.rb
#
#  Description:
#  This test suite verifies the behavior of the wget.rb script.
#  It ensures proper argument handling and mocks download logic.
#
#  Author: id774 (More info: http://id774.net)
#  Source Code: https://github.com/id774/scripts
#  License: The GPL version 3, or LGPL version 3 (Dual License).
#  Contact: idnanashi@gmail.com
#
#  Test Cases:
#    - Shows usage and exits with code 0 when invoked with the -h option.
#    - Exits with usage message when no arguments are given.
#    - Simulates downloading a file by mocking URI.open without network access.
#    - Writes downloaded content to a local file using mocked File.open.
#    - Ensures correct filename derivation from the given URL.
#
#  Dependencies:
#  This test script requires the RSpec library. To install RSpec, run:
#      gem install rspec
#
#  Running the tests:
#  Execute the test script from the command line:
#      rspec test/wget_test.rb
#
#  Version History:
#  v1.0 2025-06-24
#      Initial release. Covers usage error and mocked download test.
#
########################################################################

require 'rspec'
require 'stringio'

describe 'wget.rb' do
  let(:script_path) { File.expand_path('../../wget.rb', __FILE__) }

  before(:each) do
    @original_argv = ARGV.dup
    @original_stdout = $stdout
    $stdout = StringIO.new
  end

  after(:each) do
    ARGV.replace(@original_argv)
    $stdout = @original_stdout
  end

  it 'shows usage when -h option is given' do
    output = `ruby #{script_path} -h`
    expect($?.exitstatus).to eq(0)
    expect(output).to include('Usage')
  end

  it 'exits with usage when no arguments are given' do
    ARGV.clear
    load script_path
    expect { main }.to raise_error(SystemExit)
    output = $stdout.string
    expect(output).to include('wget.rb')
    expect(output).to include('Usage:')
  end

  it 'downloads and writes file content correctly (mocked)' do
    url = 'http://example.com/mockfile.txt'
    dummy_content = 'dummy file content'
    ARGV.replace([url])

    # Mocking URI.open to yield a StringIO with dummy content
    mock_source = double('source')
    allow(mock_source).to receive(:read).and_return(dummy_content)
    if URI.respond_to?(:open)
      allow(URI).to receive(:open).with(url).and_yield(mock_source)
    else
      allow(Kernel).to receive(:open).with(url).and_yield(mock_source)
    end

    # Mocking File.open to yield a file-like object
    mock_output = double('output')
    expect(mock_output).to receive(:print).with(dummy_content)
    allow(File).to receive(:open).with('mockfile.txt', 'w+b').and_yield(mock_output)

    load script_path
    main
  end
end

RSpec.configure do |config|
  config.color = true
  config.formatter = :progress
end
